server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -63,8 +63,17 @@ protected void unpromotableShardOperation(
         UnpromotableShardRefreshRequest request,
         ActionListener<ActionResponse.Empty> responseListener
     ) {
+        // In edge cases, the search shard may still in the process of being created when a refresh request arrives.
+        // We simply respond OK to the request because when the search shard recovers later it will use the latest
+        // commit from the proper indexing shard.
+        final var indexService = indicesService.indexService(request.shardId().getIndex());
+        if (indexService == null || indexService.hasShard(request.shardId().id()) == false) {
+            responseListener.onResponse(ActionResponse.Empty.INSTANCE);
+            return;
+        }
+
         ActionListener.run(responseListener, listener -> {
-            IndexShard shard = indicesService.indexServiceSafe(request.shardId().getIndex()).getShard(request.shardId().id());
+            IndexShard shard = indexService.getShard(request.shardId().id());
```

Comment:
I wonder if we should get the shard above instead, to ensure it is still there rather than have 2 checks. It is not the case you solved for, but I suppose the opposite could potentially happen too - the index shard lags cluster state update wise and the search shard moved on already. Rather than fail here, it seems appropriate to just say "OK" if the shard is not here (with the thinking that if it ever reappears it will use fresh data). I am not sure it is a case, but getting the shard index once using `getShardOrNull` seems simpler to reason about.